<a href="https://colab.research.google.com/github/SKM1919/liftlens/blob/main/notebooks/01_hillstrom_exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install scikit-uplift -q

import pandas as pd
from sklift.datasets import fetch_hillstrom

bunch = fetch_hillstrom(target_col='all')
df = pd.concat([bunch.data, bunch.target, bunch.treatment], axis=1)

print(df.shape)
df.head()



   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.1/42.1 kB 1.5 MB/s eta 0:00:00


Hillstrom dataset:   0%|          | 0.00/443k [00:00<?, ?iB/s]

(64000, 12)


,recency,history_segment,history,mens,womens,zip_code,newbie,channel,visit,conversion,spend,segment
0,10,2) $100 - $200,142.44,1,0,Surburban,0,Phone,0,0,0.0,Womens E-Mail
1,6,3) $200 - $350,329.08,1,1,Rural,1,Web,0,0,0.0,No E-Mail
2,7,2) $100 - $200,180.65,0,1,Surburban,1,Web,0,0,0.0,Womens E-Mail
3,9,5) $500 - $750,675.83,1,0,Rural,1,Web,0,0,0.0,Mens E-Mail
4,2,1) $0 - $100,45.34,1,0,Urban,0,Web,0,0,0.0,Womens E-Mail


In [2]:
print(df['segment'].value_counts())

df.groupby('segment')[['visit', 'conversion', 'spend']].mean()

segment
Womens E-Mail    21387
Mens E-Mail      21307
No E-Mail        21306
Name: count, dtype: int64


,visit,conversion,spend
segment,,,
Mens E-Mail,0.182757,0.012531,1.422617
No E-Mail,0.106167,0.005726,0.652789
Womens E-Mail,0.151400,0.008837,1.077202


## Key findings: Hillstrom email experiment

- **Randomisation check:** The three groups are almost equal in size (~21,300 each), so the random split worked and the comparison is fair.
- **Baseline:** Without any email, 10.6% of customers visited the site and 0.57% made a purchase.
- **Mens E-Mail** had the biggest effect: +7.7 percentage points in visits, +0.68 pp in conversions, and +$0.77 spend per customer. For every 1,000 customers emailed, that's about 77 extra visits, 7 extra purchases, and $770 extra revenue.
- **Womens E-Mail** also worked, but less: +4.5 pp visits, +0.31 pp conversions, +$0.42 spend per customer.

**Business takeaway:** Both emails drive real incremental revenue on average, and Mens performs best. But 10.6% of customers visit even without an email, so emailing everyone wastes budget on people who would have come anyway. Next step: test whether these lifts are statistically significant, then find *which* customers the email actually persuades.